File: tests/cases/science_club_lobby.py
"""Oracle-diff cases for poketcg/src/scripts/science_club_lobby.asm."""

POISON = {"a": 0xAA, "f": 0xF0, "b": 0xBB, "c": 0xCC,
          "d": 0xDD, "e": 0xEE, "hl": 0x1234}

CONTRACT = {}
CASES = {}

# >>> factory-cases-statics
wDuelResult = 0xD0C3
wNPCDuelist = 0xD0C4
wTempNPC = 0xD3AB
wLoadedNPCs = 0xD34A
wLoadedNPCTempIndex = 0xD3AA
wScriptNPC = 0xD3B6
wPlayerDirection = 0xD334
wOverworldNPCFlags = 0xD0C1
wNextScript = 0xD0C6
wOverworldMode = 0xD0BF
# <<< factory-cases-statics

# >>> factory ScienceClubLobbyAfterDuel
CONTRACT["ScienceClubLobbyAfterDuel"] = {"compare": ("a", "f", "b", "c", "d", "e", "hl"), "preserve": (), "wram_out": True}
CASES["ScienceClubLobbyAfterDuel"] = [
    {"wram": {wDuelResult: b"\x01", wNPCDuelist: b"\x99"}},
    {"wram": {
        wDuelResult: b"\x00",
        wNPCDuelist: b"\x04",
        wLoadedNPCs: b"\x04" + b"\x00" * 95,
        wLoadedNPCTempIndex: b"\xEE",
        wScriptNPC: b"\xAA",
        wPlayerDirection: b"\x01",
        wOverworldNPCFlags: b"\x00",
        wNextScript: b"\xFF\xFF",
        wOverworldMode: b"\x00",
    }, "expect": {wTempNPC: b"\x04", wScriptNPC: b"\x00", wOverworldMode: b"\x03"}},
    dict(POISON, wram={wDuelResult: b"\x01", wNPCDuelist: b"\x99"}),
]
# <<< factory ScienceClubLobbyAfterDuel

from tests.cases._schema_migration import legacy_to_schema

# >>> factory Script_Specs2
# science_club_lobby.asm:68-81 is 24 bytes of straight-line code with a single
# exit, then `rst $20` at $6BDD begins the script bytecode; completion is
# declared pre-ret there. Running past the rst enters the RST20 bytecode
# interpreter, which needs ambient scene state a probed call cannot supply.
#
# `and %11` indexes the four-entry card-id table at Data_ebe7 (03:6BE7) =
# BD BB 27 2B, so driving the RNG to all four values exercises every entry and
# produces four distinct card names. UpdateRNGSources computes
#   new_r1 = ((ctr ^ r1) << 1) | (((r2 >> 6) ^ r1) & 1)
# hence new_r1 & 3 = feedback | ((ctr ^ r1) & 1) << 1, which the seeds below
# pick directly. Index 0 must be c=3: narrowing `& 0x03` to `& 0x01` is
# invisible at c=0 and c=1.
_SPECS2_TXRAM2 = 0xCE3F
_SPECS2_RNG = 0xCACA  # wRNG1, wRNG2, wRNGCounter
CONTRACT["Script_Specs2"] = {"compare": ("a", "b", "c", "d", "e", "hl"), "preserve": ()}
CASES["Script_Specs2"] = [
    # c=3 -- table entry 0x2B; must be index 0 so the mask mutation reds
    {"a": 0, "f": 0, "b": 0, "c": 0, "d": 0, "e": 0, "hl": 0,
     "wram": {_SPECS2_TXRAM2: b"\xff\xff\xff", _SPECS2_RNG: b"\x00\x40\x01"},
     "read": {_SPECS2_TXRAM2: 3, _SPECS2_RNG: 3}},
    # c=0 -- table entry 0xBD
    {"a": 0, "f": 0, "b": 0, "c": 0, "d": 0, "e": 0, "hl": 0,
     "wram": {_SPECS2_TXRAM2: b"\xff\xff\xff", _SPECS2_RNG: b"\x00\x00\x00"},
     "read": {_SPECS2_TXRAM2: 3, _SPECS2_RNG: 3}},
    # c=1 -- table entry 0xBB
    {"a": 0, "f": 0, "b": 0, "c": 0, "d": 0, "e": 0, "hl": 0,
     "wram": {_SPECS2_TXRAM2: b"\xff\xff\xff", _SPECS2_RNG: b"\x00\x40\x00"},
     "read": {_SPECS2_TXRAM2: 3, _SPECS2_RNG: 3}},
    # c=2 -- table entry 0x27
    {"a": 0, "f": 0, "b": 0, "c": 0, "d": 0, "e": 0, "hl": 0,
     "wram": {_SPECS2_TXRAM2: b"\xff\xff\xff", _SPECS2_RNG: b"\x00\x00\x01"},
     "read": {_SPECS2_TXRAM2: 3, _SPECS2_RNG: 3}},
    dict(POISON,
         wram={_SPECS2_TXRAM2: b"\xff\xff\xff", _SPECS2_RNG: b"\x00\x40\x01"},
         read={_SPECS2_TXRAM2: 3, _SPECS2_RNG: 3}),
]
# <<< factory Script_Specs2

SCHEMA2_CASES = legacy_to_schema(CASES, CONTRACT)
# >>> factory-completion Script_Specs2
# The routine ends at its `rst $20`; legacy_to_schema always emits "return".
for _rec in SCHEMA2_CASES["Script_Specs2"]:
    _rec["completion"] = {"mode": "pre-ret", "pc": 0x6BDD}
# <<< factory-completion Script_Specs2

MUTATIONS = {}
# >>> factory-mutation ScienceClubLobbyAfterDuel
MUTATIONS["ScienceClubLobbyAfterDuel"] = {"source_symbol": "ScienceClubLobbyAfterDuel", "before": "	FindEndOfDuelScriptResult r = FindEndOfDuelScript(ScienceClubLobbyAfterDuelTable);", "after": "	FindEndOfDuelScriptResult r = FindEndOfDuelScript((uint16_t)(ScienceClubLobbyAfterDuelTable + 1u));", "case_ids": ["ScienceClubLobbyAfterDuel-0"]}
# <<< factory-mutation ScienceClubLobbyAfterDuel
# >>> factory-mutation Script_Specs2
MUTATIONS["Script_Specs2"] = {"source_symbol": "Script_Specs2", "before": "\tuint8_t c = (uint8_t)(rng & 0x03u);", "after": "\tuint8_t c = (uint8_t)(rng & 0x01u);", "case_ids": ["Script_Specs2-0", "Script_Specs2-3", "Script_Specs2-4"]}
# <<< factory-mutation Script_Specs2

File: src/home/science_club_lobby.c
#include "home/science_club_lobby.h"

#include "generated/hram.h"
#include "generated/wram.h"
#include "mem.h"
/* >>> factory statics */
#define DATA_EBE7 0x6BE7u
#define DATA_EBE7_BANK 3u
#include "home/science_club_lobby.h"
#include "home/grass_club_entrance.h"
#include "generated/wram.h"
#define ScienceClubLobbyAfterDuelTable 0x6b5eu
#include "home/scripting.h"
#define EVENT_TEMP_DUELED_IMAKUNI 0x03u
#define EVENT_IMAKUNI_STATE 0x13u
#define EVENT_IMAKUNI_ROOM 0x34u
#define IMAKUNI_TALKED 0x02u
#define IMAKUNI_SCIENCE_CLUB 0x01u
#define MUSIC_IMAKUNI 0x10u
/* <<< factory statics */

/* >>> factory ScienceClubLobbyAfterDuel */
ScienceClubLobbyAfterDuelResult ScienceClubLobbyAfterDuel(void)
{
	gb_write8(0x2000u, 0x03u);
	FindEndOfDuelScriptResult r = FindEndOfDuelScript(ScienceClubLobbyAfterDuelTable);
	return (ScienceClubLobbyAfterDuelResult){r.a, r.f, r.b, r.c, r.d, r.e, r.hl};
}
/* <<< factory ScienceClubLobbyAfterDuel */

/* >>> factory Script_Specs2 */
/* science_club_lobby.asm:68-81 -- the routine's entire CODE portion, 24 bytes,
 * straight-line with a single exit. The `start_script` macro assembles to
 * `rst $20` at $6BDD, where the code ends and the script bytecode begins, so
 * the cases declare completion pre-ret there.
 *
 * `and %11` keeps the low two bits of UpdateRNGSources' return, indexing a
 * four-entry card-id table at Data_ebe7 (03:6BE7). GetCardName returns the new
 * `de`, which is stored little-endian to wTxRam2 by `ld [hli],a` / `ld [hl],d`,
 * leaving hl at wTxRam2+1. `d` is zeroed before the call, so only `e` carries
 * the table byte in. */
ScriptSpecs2Result Script_Specs2(void)
{
	uint8_t rng = UpdateRNGSources();
	uint8_t c = (uint8_t)(rng & 0x03u);
	uint16_t hl = (uint16_t)(DATA_EBE7 + c);
	uint8_t e = rom_ptr(DATA_EBE7_BANK, hl)[0];
	uint16_t de = GetCardName(e);
	hl = wTxRam2_ADDR;
	gb_write8(hl, (uint8_t)(de & 0xFFu));
	hl = (uint16_t)(hl + 1u);
	gb_write8(hl, (uint8_t)(de >> 8));
	return (ScriptSpecs2Result){(uint8_t)(de & 0xFFu), 0u, c,
	                            (uint8_t)(de >> 8), (uint8_t)(de & 0xFFu), hl};
}
/* <<< factory Script_Specs2 */

/* >>> factory Preload_ImakuniInScienceClubLobby */
/* science_club_lobby.asm:13-30: every `.dont_load` exit is `or a` on the value last
 * loaded (the state, the temp-duel flag or the room); the load exit is `scf`
 * over the `cp`'s Z. */
PreloadImakuniInScienceClubLobbyResult Preload_ImakuniInScienceClubLobby(void)
{
	uint8_t state = GetEventValue(EVENT_IMAKUNI_STATE);
	if (state < IMAKUNI_TALKED)
		return (PreloadImakuniInScienceClubLobbyResult){state, state == 0u ? 0x80u : 0u};
	uint8_t dueled = GetEventValue(EVENT_TEMP_DUELED_IMAKUNI);
	if (dueled != 0u)
		return (PreloadImakuniInScienceClubLobbyResult){dueled, 0u};
	uint8_t room = GetEventValue(EVENT_IMAKUNI_ROOM);
	if (room != IMAKUNI_SCIENCE_CLUB)
		return (PreloadImakuniInScienceClubLobbyResult){room, room == 0u ? 0x80u : 0u};
	wDefaultSong = MUSIC_IMAKUNI;
	return (PreloadImakuniInScienceClubLobbyResult){MUSIC_IMAKUNI, 0x90u};
}
/* <<< factory Preload_ImakuniInScienceClubLobby */
